run_scans.py

```diff
@@ -661,6 +662,7 @@ def ultraThreshold(args):
     parser_sbitThresh.add_argument("--scanmin",type=int,default=0,help="Minimum CFG_THR_ARM_DAC")
     parser_sbitThresh.add_argument("--scanmax",type=int,default=255,help="Maximum CFG_THR_ARM_DAC")
     parser_sbitThresh.add_argument("--stepSize",type=int,default=1,help="Step size to use when scanning CFG_THR_ARM_DAC")
+    parser_sbitThresh.add_argument("--waitTime",type=int,default=1000,help="Length of the time window within which the rate is measured, in milliseconds")
```

Comment:
Unless there is inertia elsewhere in the existing codebase for milleseconds, I would discretize this in seconds

Reply:
Do we want to allow fractional values of seconds? I believe we are not sending any `float`s in the rpc calls and it is not clear to me whether it is possible or a good idea. But we could make the argument a `float` in seconds in this pull request and then convert it to an `int` in milliseconds for the `cmsgemos` function.

Reply:
I suppose we could... but I think we're not interested in going below 1s.
Where it could be interesting would be if the scan itself was much more granular and could take multiple measurements to get proper uncertainties, but I think this is also not the most important thing.
What we should get in the immediate term is the ability to use the FW counters for some programmable time and determine the rate in Hz from the counter and the integration time (this last bit may only make sense to do in the analysis, since everything up to that point is just an integer type, right?)

In the CTP7 FW, the rate calculator is hard-coded to rearm after 1s, though the counter itself could be read whenever and just rolls over.
In the OH FW, the counter integrates for a programmable time and then resets

Reply:
> I suppose we could... but I think we're not interested in going below 1s.
> Where it could be interesting would be if the scan itself was much more granular and could take multiple measurements to get proper uncertainties, but I think this is also not the most important thing.
> What we should get in the immediate term is the ability to use the FW counters for some programmable time and determine the rate in Hz from the counter and the integration time (this last bit may only make sense to do in the analysis, since everything up to that point is just an integer type, right?)

So, you are proposing that instead of storing a `rate` in the output tree, which is what we are doing now, we store both the `count` and the `waitTime`?

> 
> In the CTP7 FW, the rate calculator is hard-coded to rearm after 1s, though the counter itself could be read whenever and just rolls over.
> In the OH FW, the counter integrates for a programmable time and then resets

OK, I see now that we are measuring an all-VFAT rate and then 24 per-VFAT rates. It seems that the register used in the all-VFAT rate measurement ("GEM_AMC.TRIGGER.OH0.TRIGGER_RATE") is on the CTP7 while the register used in per-VFAT rate measurement ("GEM_AMC.OH.OH0.FPGA.TRIG.CNT.VFAT0_SBITS") is on the OH.

Is the register "GEM_AMC.TRIGGER.OH0.TRIGGER_RATE" actually a counter or rate?

Reply:
> > I suppose we could... but I think we're not interested in going below 1s.
> > Where it could be interesting would be if the scan itself was much more granular and could take multiple measurements to get proper uncertainties, but I think this is also not the most important thing.
> > What we should get in the immediate term is the ability to use the FW counters for some programmable time and determine the rate in Hz from the counter and the integration time (this last bit may only make sense to do in the analysis, since everything up to that point is just an integer type, right?)
> 
> So, you are proposing that instead of storing a `rate` in the output tree, which is what we are doing now, we store both the `count` and the `waitTime`?
> 

Hmm, no, I'm proposing that we store the same thing, though I was thinking that we were storing the count as the rate, since it was integrated over 1s.
If the count was previously being divided by some wait time (of 1s) and this value being stored, then we should continue, except with the now configurable wait time.

> > In the CTP7 FW, the rate calculator is hard-coded to rearm after 1s, though the counter itself could be read whenever and just rolls over.
> > In the OH FW, the counter integrates for a programmable time and then resets
> 
> OK, I see now that we are measuring an all-VFAT rate and then 24 per-VFAT rates. It seems that the register used in the all-VFAT rate measurement ("GEM_AMC.TRIGGER.OH0.TRIGGER_RATE") is on the CTP7 while the register used in per-VFAT rate measurement ("GEM_AMC.OH.OH0.FPGA.TRIG.CNT.VFAT0_SBITS") is on the OH.
> 
> Is the register "GEM_AMC.TRIGGER.OH0.TRIGGER_RATE" actually a counter or rate?

This register is truly a rate, and it has an integration time of 1s, not programmable, though this is not as important, one could read this rate every 1+ second to account for delays, and then take the average to get the time averaged rate, but again, this particular value is less important than the per-VFAT value.
Alternatively, we can do the same thing as is done for each VFAT, since there is a CTP7 `OHX_TRIGGER_CNT` register that behaves (mostly) the same as the per-VFAT counters in the OH.
In the interest of the minimal change now, let's only focus on the per-VFAT rate in this set of PRs

Reply:
ok, so the `waitTime` will be applied only to the per-VFAT counters

Reply:
The count was previously not being divided by anything, which is equivalent to dividing by 1 s. Now, I think what we want to do is divide by the `waitTime`, which I have done here:

https://github.com/AndrewLevin/ctp7_modules/blob/sbitrate_waittime/src/calibration_routines.cpp#L711

Reply:
Which was probably not a good idea since the right-hand side is a float while the left-hand side is an integer. Maybe it is a better idea to return the count, and then perform the division in vfatqc-python-scripts.

Reply:
> Which was probably not a good idea since the right-hand side is a float while the left-hand side is an integer. Maybe it is a better idea to return the count, and then perform the division in vfatqc-python-scripts.

Yeah, this was what I was thinking was the case.   Returning the count, then let the thing that fills the tree take care of converting to the appropriate rate.

Reply:
I have changed all of the 4 pull requests to use seconds instead of milliseconds now.